torchtune/modules/transformer.py

```diff
@@ -61,23 +103,126 @@ def forward(
         Returns:
             Tensor: output tensor with same shape as input
                 [batch_size x seq_length x embed_dim]
+        """
+        # Input tensor and attention output have the same shape
+        # [b, s, d]
+        # Norm applied before self-attention
+        attn_out = self.attn(self.attn_norm(x), mask=mask, input_pos=input_pos)
+
+        # Residual connection; shape: [batch_size, seq_length, embed_dim]
+        h = self.attn_scale(attn_out) + x
+
+        # Norm applied before the feedforward layer
+        mlp_out = self.mlp(self.mlp_norm(h))
+
+        # Residual connection; shape: [batch_size, seq_length, embed_dim]
+        out = h + self.mlp_scale(mlp_out)
+        return out
+
+
+class TransformerCrossAttentionLayer(nn.Module):
```

Comment:
I think I have very fundamental misunderstanding of what this layer is doing. I thought we always have two inputs here but then I'm seeing support for when encoder_input is None. Pardon my ignorance, but can you expand on what this is doing conceptually?

Reply:
The layer does cross attention between the encoder and decoder inputs. But when there is no encoder input, it is a noop.

Reply:
I left a comment in your other PR about this: "I find it a bit misleading. I dont think that someone should call a cross attention layer if they dont have anything to cross attend. IMO, it may lead to silent errors that are hard to debug, because it wont raise any errors. If the enconder_input is None, it should probably be resolved where the crossattention is being called."